src/admin/render_readmes.rs

```diff
@@ -184,26 +186,35 @@ fn render_pkg_readme<R: Read>(mut archive: Archive<R>, pkg_name: &str) -> anyhow
     };
 
     let rendered = {
-        let readme = manifest.package().readme();
-        if !readme.is_some() {
-            return Ok("".to_string());
-        }
-
-        let readme_path = readme.as_path().unwrap_or_else(|| Path::new("README.md"));
-        let path = Path::new(pkg_name).join(readme_path);
+        let readme = manifest
+            .package
+            .as_ref()
+            .and_then(|p| p.readme.as_ref())
+            .and_then(|r| r.as_ref().as_local());
+
+        let readme_path = match readme {
+            Some(StringOrBool::Bool(bool)) if !(*bool) => return Ok("".to_string()),
+            Some(StringOrBool::String(path)) => PathBuf::from(path),
+            _ => PathBuf::from("README.md"),
+        };
+
+        let path = Path::new(pkg_name).join(&readme_path);
         let contents = find_file_by_path(&mut entries, Path::new(&path))
             .with_context(|| format!("Failed to read {} file", readme_path.display()))?;
 
         // pkg_path_in_vcs Unsupported from admin::render_readmes. See #4095
         // Would need access to cargo_vcs_info
         let pkg_path_in_vcs = None;
 
-        text_to_html(
-            &contents,
-            &readme_path,
-            manifest.package().repository(),
-            pkg_path_in_vcs,
-        )
+        let package = manifest.package.unwrap();
```

Comment:
We should reproduce the comment about why this is safe here as well, since it's not guaranteed a future contributor will know the context.

Reply:
I thought I had implemented it in a way where `manifest.package == None` resulted in an early return from the function, but as it turns out I had changed the code to treat `readme == None` as "use the default readme path" which means there is no early return in this case anymore. I've changed the code to use the same `and_then()` chain as above instead.